src/server.rs

```diff
@@ -255,20 +255,62 @@ impl ServerPlugin {
 /// On deserialization mappings should be processed first, so all referenced entities after it will behave correctly.
 fn collect_mappings(
     messages: &mut ReplicationMessages,
-    entity_map: &mut ClientEntityMap,
+    archetypes_info: &ReplicatedArchetypesInfo,
+    replication_rules: &ReplicationRules,
+    world: &World,
+    change_tick: &SystemChangeTick,
 ) -> bincode::Result<()> {
-    for (message, _, client_info) in messages.iter_mut_with_info() {
-        message.start_array();
+    for (init_message, _) in messages.iter_mut() {
+        init_message.start_array();
+    }
 
-        if let Some(mappings) = entity_map.0.get_mut(&client_info.id()) {
-            for mapping in mappings.drain(..) {
-                message.write_client_mapping(&mapping)?;
+    for archetype_info in archetypes_info.iter_client_mapped() {
```

Comment:
Maybe use `Query` instead of iterating of the archetypes manually?
It's simpler and faster because it not only caches archetypes, but also uses more optimal iteration if all components are in table storage (which is the case).